homeassistant/components/vacuum/ecovacs.py

```diff
@@ -0,0 +1,222 @@
+        self.device.connect_and_wait_until_ready()
+        try:
+            self._name = '{}'.format(self.device.vacuum['nick'])
+        except KeyError:
+            # In case there is no nickname defined, use the device id
+            self._name = '{}'.format(self.device.vacuum['did'])
+
+        self._fan_speed = None
+        self._error = None
+        _LOGGER.debug("Vacuum initialized: %s", self.name)
+
+    async def async_added_to_hass(self) -> None:
+        """d."""
+        # Fire off some queries to get initial state
+        self.device.statusEvents.subscribe(self.on_status)
+        self.device.batteryEvents.subscribe(self.on_battery)
+        self.device.errorEvents.subscribe(self.on_error)
+        self.device.lifespanEvents.subscribe(self.on_lifespan)
+
+    def on_status(self, status):
+        """Handle the status of the robot changing."""
+        self.schedule_update_ha_state()
+
+    def on_battery(self, battery_level):
+        """Handle the battery level changing on the robot."""
+        self.schedule_update_ha_state()
+
+    def on_lifespan(self, lifespan):
+        """Handle component lifespan reports from the robot."""
+        self.schedule_update_ha_state()
+
+    def on_error(self, error):
+        """Handle an error event from the robot.
+
+        This will not change the entity's state. If the error caused the state
+        to change, that will come through as a separate on_status event
+        """
+        if error == 'no_error':
+            self._error = None
+        else:
+            self._error = error
+
+        self.hass.bus.fire('ecovacs_error', {
+            'entity_id': self.entity_id,
+            'error': error
+        })
+        self.schedule_update_ha_state()
+
+    @property
+    def should_poll(self) -> bool:
+        """Return True if entity has to be polled for state."""
+        return False
+
+    @property
+    def unique_id(self) -> str:
+        """Return an unique ID."""
+        return self.device.vacuum.get('did', None)
+
+    @property
+    def is_on(self):
+        """Return true if vacuum is currently cleaning."""
+        return self.device.vacuum_status in STATUSES_CLEANING
+
+    @property
+    def is_charging(self):
+        """Return true if vacuum is currently charging."""
+        return self.device.vacuum_status in STATUSES_CHARGING
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Flag vacuum cleaner robot features that are supported."""
+        return SUPPORT_ECOVACS
+
+    @property
+    def status(self):
+        """Return the status of the vacuum cleaner."""
+        return self.device.vacuum_status
+
+    def return_to_base(self, **kwargs):
+        """Set the vacuum cleaner to return to the dock."""
+        from sucks import Charge
+        self.device.run(Charge())
+
+    @property
+    def battery_icon(self):
```

Comment:
This seems a flaw in the vacuum ABC if you need to implement this yourself. Your implementation seems like how it should be?

Reply:
It definitely is an issue with the ABC having no formal support for determining if the bot is charging. The reason I chose not to fix the ABC here is because the proper fix is the rearchitecture discussed here: https://github.com/home-assistant/architecture/issues/29

I plan on updating this component after/alongside @cnrd makes the foundational changes